subsys/net/lib/download_client/src/download_client.c

```diff
@@ -52,40 +49,6 @@ static const char *str_family(int family)
 	}
 }
 
-static int socket_timeout_set(int fd, int type)
-{
-	int err;
-	uint32_t timeout_ms;
-
-	if (type == SOCK_STREAM) {
-		timeout_ms = CONFIG_DOWNLOAD_CLIENT_TCP_SOCK_TIMEO_MS;
-	} else {
-		timeout_ms = CONFIG_DOWNLOAD_CLIENT_UDP_SOCK_TIMEO_MS;
-	}
-
-	if (timeout_ms <= 0) {
-		return 0;
-	}
-
-	struct timeval timeo = {
-		.tv_sec = (timeout_ms / 1000),
-		.tv_usec = (timeout_ms % 1000) * 1000,
-	};
-
-#if defined(CONFIG_POSIX_API)
-	LOG_INF("Configuring socket timeout (%d s)", (int32_t)timeo.tv_sec);
-#else
-	LOG_INF("Configuring socket timeout (%lld s)", timeo.tv_sec);
-#endif
-	err = setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeo, sizeof(timeo));
```

Comment:
The initial intention behind using poll was that poll allows a variable timeout. This is in particular needed to implement the CoAP backoff mechanism. Of course, you could also change the socket timeout via 'SO_SNDTIMEO' and `SO_RCVTIMEO ` via `setsockopt` but this is not really a good code style that I have tried to avoid.

My idea behind using polling was: Can we send anything and should be something sent (POLLOUT)? Are we expecting incoming data (POLLIN)? If no, in both cases a poll timeout is raised and proper actions are done e.g. pending_cycle is called for coap. For http there is no mechanism such as coap exponential backoff implemented. The code relies on that at some point there is a http error/close in case of network issues.

What makes my approach better: By using POLLIN/POLLOUT we never read/write to socket that that does not have any data for us (read) or is not ready (write). This makes in my opinion the code quite straight forward. If poll times out, we use a retry mechanism such as coap exponential backoff. Otherwise depending on how poll events are set, we do necessary actions such as requesting a new block or reading from socket.

I agree, I had to split the commit into two separate patches. One that introduces poll and one that introduces CoAP backoff. If you wish so, I can do that :)

Reply:
> Of course, you could also change the socket timeout via 'SO_SNDTIMEO' and SO_RCVTIMEO via setsockopt but this is not really a good coding practice that I have tried to avoid.

Why do you think it's a bad practice to modify timeouts set on a socket with `setsockopt()`? I think the contrary, that getting `poll()` involved in a case when a single socket is used is an overhead. IMO it just adds additional complexity with no real benefit (for instance with additional error handling needed).

> What makes my approach better: By using POLLIN/POLLOUT we never read/write to socket that that does not have any data for us (read) or is not ready (write).

But that's the whole point of using blocking sockets - the application does not need to worry about whether there is new data available, the `recv()` will block until there is (or timeout occurs in our case). Same goes for `send()`.

> I agree, I had to split the commit into two separate patches. One that introduces poll and one that introduces CoAP backoff. If you wish so, I can do that :)

Please note, that we use `download_client` library massively in NCS - it's really hard for us to justify significant changes to the library just for the sake of refactoring.

Reply:
If possible, I suggest to split this patch and work on adding CoAP backoff support with blocking sockets (with a timeout) first. I think the patch will be smaller and it'll be easier to review and merge.

Reply:
> Please note, that we use download_client library massively in NCS 

But obviously not a lot for COAP, as is evident by #6997 and #6788, which both made it pretty much unusable (which is unfortunate since the lwm2m fota implementation switched to it). (And, to a lesser, but still important extent the issue this PR aims to resolve: right now, when the server never responds (i.e. because it doesn't exist) the download process will just resend requests indefinitely).

> But that's the whole point of using blocking sockets - the application does not need to worry about whether there is new data available, the recv() will block until there is (or timeout occurs in our case). Same goes for send().

I think this argument is along a similar line: Yes, for TCP/HTTP, blocking sockets (and even no "proper" timeout handling) are acceptable since the underlying transport protocol takes care of that stuff, but with COAP the communication is asynchronous in nature and proper care needs to be taken. IMHO poll is the best way to address this:
Take for the example the case that the initial receive times out. How long should we wait for the send operation to finish before cycling the coap retransmit counter? A naive blocking implementation wouldn't change the timeout for the send call, ultimately using the wrong timeouts for the entire coap transmission. Unifying the blocking/waiting behind one single call and having to only care about timeout calculation at one single point simplifies this and is imho easier that a correct implementation using blocking recv/send calls.

> it's really hard for us to justify significant changes to the library just for the sake of refactoring

We also rely on this library heavily, so we share your concerns about it's quality. In fact @chmielewskiandreas is currently looking into the possibility of adding unit tests to it.

> If possible, I suggest to split this patch and work on adding CoAP backoff support with blocking sockets (with a timeout) first

I believe a correct blocking recv/send implementation is more complicated, so I would suggest to go with @chmielewskiandreas suggestions of splitting the commits in the poll and timeout implementation and then evaluating them separately.

Reply:
> But obviously not a lot for COAP

That is true. But I still think that we can focus on fixing issues with CoAP part, w/o refactoring half of the library.

> A naive blocking implementation wouldn't change the timeout for the send call, ultimately using the wrong timeouts for the entire coap transmission.

Let's make it non-naive then. `struct coap_pending` provide `t0` and `timeout` values - this information is enough to determine when should the next retransmission take place, and thus what timeout should be set on a socket.

> Unifying the blocking/waiting behind one single call and having to only care about timeout calculation at one single point simplifies this and is imho easier that a correct implementation using blocking recv/send calls.

Still, the current implementation does not take into account time spent in `poll()` when `POLLOUT` is set - so in worst-case scenario the retransmission timeout will be 2x expected timeout.

Reply:
@rlubos: I have implemented CoAP backoff mechanism in a way as you have proposed. There are no fundamental changes for your http client.

Reply:
https://github.com/nrfconnect/sdk-nrf/pull/7288